From 0d7b69f3ef9679cee1c0d98ec5926641915c291c Mon Sep 17 00:00:00 2001
From: "Liu,Dalei" <daleix.liu@intel.com>
Date: Wed, 30 Dec 2015 16:35:43 +0800
Subject: [PATCH 4/5] atomisp: [WORKAROUND] Hardcoded 5MP crop for BTNS

On Brightonshores, when customer require 5M(2560x1920)
resolution frame. We didn't have an appropriate settings
from ov8858. And BTNS is using ISP2.7 still pipe which
dosen't support downscaling. Therefore we need this
workaroud.
Plese also note, this patch will be reverted when we have
a proper 5MP cropped sensor setting.

Change-Id: Ib010a8a098406b6904fa092a090aa42f9405d2d4
Original-Tracked-On: #H1504133721
Signed-off-by: Liu,Dalei <daleix.liu@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/92954
Reviewed-by: Bin B Han <bin.b.han@intel.com>
Approver: Bin B Han <bin.b.han@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Build: Dessert M BXT RVP (SC)
Build: Dessert M CHT FFD (SC)
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-15122
Signed-off-by: Yang Fan F <fan.f.yang@intel.com>
Reviewed-on: https://android.intel.com:443/455606
---
 .../media/pci/atomisp2/atomisp_driver/atomisp_cmd.c | 15 +++++++++++++++
 .../atomisp2/atomisp_driver/atomisp_compat_css20.c  | 21 +++++++++++++++++++++
 2 files changed, 36 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
index 458682c..e899fd1 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_cmd.c
@@ -89,6 +89,13 @@ union host {
 	} ptr;
 };
 
+#ifdef CONFIG_PLATFORM_BTNS
+/* This flag is for 5M(2560x1920) request from customer.
+ * It's a workaround method to conform FW request because
+ * we didn't have appropriate resolutiong settings */
+int btns_5mp_crop_workaround = 0;
+#endif
+
 /*
  * atomisp_kernel_malloc: chooses whether kmalloc() or vmalloc() is preferable.
  *
@@ -5612,6 +5619,14 @@ int atomisp_set_fmt(struct video_device *vdev, struct v4l2_format *f)
 			|| r->height > f->fmt.pix.height))
 			dev_warn(isp->dev,
 				 "Main Resolution config smaller then Vf Resolution. Force to be equal with Vf Resolution.");
+#ifdef CONFIG_PLATFORM_BTNS
+	/* When customer need capture 2560x1920, we set flag to workaround */
+		if ((f->fmt.pix.width == 2560) &&
+				(f->fmt.pix.height == 1920))
+			btns_5mp_crop_workaround = 1;
+		else
+			btns_5mp_crop_workaround = 0;
+#endif
 	}
 
 	/* Pipeline configuration done through subdevs. Bail out now. */
diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
index ce929f3..1c1d193 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_compat_css20.c
@@ -72,6 +72,13 @@ struct bayer_ds_factor {
 	unsigned int denominator;
 };
 
+#ifdef CONFIG_PLATFORM_BTNS
+#define BTNS_CROP_PADDING_WIDTH		128
+#define BTNS_CROP_PADDING_HEIGHT	96
+
+extern int btns_5mp_crop_workaround;
+#endif
+
 void atomisp_css_debug_dump_sp_sw_debug_info(void)
 {
 	ia_css_debug_dump_sp_sw_debug_info();
@@ -2008,8 +2015,22 @@ int atomisp_css_input_set_effective_resolution(
 {
 	struct ia_css_stream_config *s_config =
 			&asd->stream_env[stream_id].stream_config;
+
+#ifdef CONFIG_PLATFORM_BTNS
+	if (btns_5mp_crop_workaround == 1) {
+		s_config->input_config.effective_res.width =
+			2560 + BTNS_CROP_PADDING_WIDTH;
+		s_config->input_config.effective_res.height =
+			1920 + BTNS_CROP_PADDING_HEIGHT;
+	} else {
+		s_config->input_config.effective_res.width = width;
+		s_config->input_config.effective_res.height = height;
+	}
+#else
 	s_config->input_config.effective_res.width = width;
 	s_config->input_config.effective_res.height = height;
+#endif
+
 	return 0;
 }
 
-- 
1.9.1

